osdep/als-linux.c

```diff
@@ -0,0 +1,192 @@
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "common/common.h"
+#include "common/global.h"
+#include "common/msg.h"
+#include "misc/bstr.h"
+#include "misc/mp_assert.h"
+#include "misc/path_utils.h"
+#include "osdep/als.h"
+#include "osdep/io.h"
+#include "player/core.h"
+#include "ta/ta_talloc.h"
+
+// For the sysfs API documentation, refer to
+//     Documentation/ABI/testing/sysfs-bus-iio
+// In Kernel tree.
+#define SYSFS_BASEDIR "/sys/bus/iio/devices"
+#define INP_RAW "in_illuminance_raw"
+#define INP_OFF "in_illuminance_offset"
+#define INP_SCL "in_illuminance_scale"
+
+struct mp_als {
+    DIR *device_dir;
+    struct mp_log *log;
+
+    // Flag we set to true if we've looked for IIO devices and found none. This
+    // way, we know that we shouldn't look again next time.
+    bool found_no_devices;
+};
+
+static void mp_als_destroy(void *p)
+{
+    struct mp_als *state = p;
+
+    if (state->device_dir) {
+        closedir(state->device_dir);
+    }
+}
+
+struct mp_als *mp_als_create(void *parent, MPContext *mpctx)
+{
+    struct mp_als *state =
+        talloc_zero(parent, struct mp_als);
+
+    state->log = mp_log_new(state, mpctx->global->log, "linux-als");
+
+    talloc_set_destructor(state, mp_als_destroy);
+
+    return state;
+}
+
+// Assumes that state->device_dir is NULL
+static int acquire_sensor(struct mp_als *state)
+{
+    mp_assert(!state->device_dir);
+    DIR *devices_dir = opendir(SYSFS_BASEDIR);
+    if (!devices_dir) {
+        MP_WARN(state,
+                "No IIO device directory in sysfs: %s\n",
+                mp_strerror(errno));
+        return -1;
+    }
+
+    void *tmp = talloc_new(NULL);
+
+    struct dirent *ent;
+    while ((ent = readdir(devices_dir))) {
+        // The directory also contains files named `trigger*` which we don't
+        // care about; only include ones for devices.
+        if (!bstr_startswith0(bstr0(ent->d_name), "iio:device"))
+            continue;
+
+        // Check if in_illuminance_raw exists so we know we have an ambient
+        // light sensor.
+        char *inp_raw_path = mp_path_join(tmp, ent->d_name, INP_RAW);
+
+        if (faccessat(dirfd(devices_dir), inp_raw_path, R_OK, 0) != 0) {
+            // Couldn't stat, continue
+            continue;
+        }
+
+        // We found an ALS
+
+        char *dev_path = mp_path_join(tmp, SYSFS_BASEDIR, ent->d_name);
+
+        state->device_dir = opendir(dev_path);
+        if (!state->device_dir) {
+            MP_ERR(state, "Failed to open ambient light sensor %s: %s\n",
+                   ent->d_name,
+                   mp_strerror(errno));
+            talloc_free(tmp);
+            closedir(devices_dir);
+            return -1;
+        }
```

Comment:
sorry, noticed this just now: there should be a `break` here if there are multiple ALS

Reply:
For multiple devices support, shouldn't return in `if (!state->device_dir)` be `continue` too?